### PR TITLE
[Admin][CatalogPromotion] Add filters

### PR DESCRIPTION
**Pull request by NoResponseMate**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12                   |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| Related tickets | -
| License         | MIT                                                          |

image

### CODE REVIEW

File: features/promotion/managing_catalog_promotions/catalog_promotion_filtration/filtering_catalog_promotions_by_channel.feature [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    @ui @api
+    Scenario: Filtering catalog promotions by a chosen channel
+        When I browse catalog promotions
+        And I filter by "Web-EU" channel
+        Then I should see a catalog promotion with name "Surprise sale"
+        And I should see a catalog promotion with name "Spring sale"
+        But I should not see a catalog promotion with name "Winter sale"
+
+    @ui @api
+    Scenario: Filtering catalog promotions by another channel
```
Comment (Zales0123) [inline features/promotion/managing_catalog_promotions/catalog_promotion_filtration/filtering_catalog_promotions_by_channel.feature]: In theory, this second scenario ensures us the feature works, but generally it's describing exactly the same feature as the previous one, so I think we can safely remove it 🖖

File: src/Sylius/Behat/Context/Setup/CatalogPromotionContext.php [hunk 2]
```diff
@@ -756,13 +758,34 @@ public function theCatalogPromotionPriorityIs(CatalogPromotionInterface $catalog
...
+     * @Given the processing of :catalogPromotion has failed
+     */
+    public function theProcessingOfCatalogPromotionHasFailed(CatalogPromotionInterface $catalogPromotion): void
+    {
+        $this->setCatalogPromotionState($catalogPromotion, CatalogPromotionStates::STATE_FAILED);
+    }
+
+    private function setCatalogPromotionState(CatalogPromotionInterface $catalogPromotion, string $state): void
+    {
+        $catalogPromotion->setState($state);
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/CatalogPromotionContext.php]: Please, no :( If we manipulate catalog promotions' states, we should definitely use state machine transitions to achieve them. I know, we have such abomination already in `OrderContext`, but we should definitely get rid of them 🖖

File: src/Sylius/Behat/Context/Ui/Admin/ManagingCatalogPromotionsContext.php [hunk 3]
```diff
@@ -593,6 +595,87 @@ public function iEditItToHaveEmptyFixedDiscountInChannel(ChannelInterface $chann
...
+     */
+    public function iFilterByState(string $state): void
+    {
+        $this->filterElement->chooseState(ucfirst($state));
+        $this->filterElement->filter();
+    }
+
+    /**
+     * @When I search by :phrase name
+     * @When I search by :phrase code
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingCatalogPromotionsContext.php:627]: I wonder isn't `I search by :phrase` enough 🤔

File: src/Sylius/Bundle/AdminBundle/Resources/config/grids/catalog_promotion.yaml [hunk 4]
```diff
@@ -46,6 +46,43 @@ sylius_grid:
...
+                    label: sylius.ui.enabled
+                state:
+                    type: select
+                    label: sylius.ui.state
+                    form_options:
+                        choices:
+                            sylius.ui.active: active
+                            sylius.ui.failed: failed
+                            sylius.ui.inactive: inactive
+                            sylius.ui.processing: processing
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/config/grids/catalog_promotion.yaml]: We could use constants from `Sylius\Component\Promotion\Model\CatalogPromotionStates` here

File: src/Sylius/Behat/Context/Ui/Admin/ManagingCatalogPromotionsContext.php (hunk 3 above)
Comment (NoResponseMate) [reply src/Sylius/Behat/Context/Ui/Admin/ManagingCatalogPromotionsContext.php:627]: No problem for UI, but in the case of API it'd necessitate choosing either or.

Comment (Zales0123) [reply src/Sylius/Behat/Context/Ui/Admin/ManagingCatalogPromotionsContext.php:627]: 👍

**Review by AdamKasp — APPROVED** [review]

**Comment by AdamKasp** [comment]
Thank you, Jan! :1st_place_medal:

**Pull request merged**